src/eth_rpc/servers/kakarot_rpc.rs

```diff
@@ -0,0 +1,96 @@
+use crate::{
+    eth_provider::{provider::EthereumProvider, starknet::kakarot_core::to_starknet_transaction},
+    eth_rpc::api::kakarot_api::KakarotApiServer,
+};
+use jsonrpsee::core::{async_trait, RpcResult};
+use jsonrpsee_types::{error::INVALID_PARAMS_CODE, ErrorObject};
+use reth_primitives::B256;
+use starknet::{
+    core::{
+        crypto::compute_hash_on_elements,
+        types::{BroadcastedInvokeTransaction, BroadcastedInvokeTransactionV1, Felt},
+    },
+    providers::Provider,
+};
+use std::convert::TryInto;
+
+#[derive(Debug)]
+pub struct KakarotRpc<EP, SP> {
+    eth_provider: EP,
+    starknet_provider: SP,
+}
+
+impl<EP, SP> KakarotRpc<EP, SP> {
+    pub const fn new(eth_provider: EP, starknet_provider: SP) -> Self {
+        Self { eth_provider, starknet_provider }
+    }
+}
+trait ToElements {
+    fn try_into_v1(self) -> Result<BroadcastedInvokeTransactionV1, eyre::Error>;
+}
+
+impl ToElements for BroadcastedInvokeTransaction {
+    fn try_into_v1(self) -> Result<BroadcastedInvokeTransactionV1, eyre::Error> {
+        match self {
+            Self::V1(tx_v1) => Ok(tx_v1),
+            Self::V3(_) => Err(eyre::eyre!("Transaction is V3, cannot convert to V1")),
+        }
+    }
+}
+
+#[async_trait]
+impl<EP, SP> KakarotApiServer for KakarotRpc<EP, SP>
+where
+    EP: EthereumProvider + Send + Sync + 'static,
+    SP: Provider + Send + Sync + 'static,
+{
+    async fn get_starknet_transaction_hash(&self, hash: B256, retries: u8) -> RpcResult<Option<B256>> {
+        // Retrieve the stored transaction from the database.
+        let transaction = self.eth_provider.transaction_by_hash(hash).await?;
+
+        if let Some(transaction) = transaction {
+            // Convert the `Transaction` instance to a `TransactionSigned` instance.
+            let transaction_signed_ec_recovered: reth_primitives::TransactionSignedEcRecovered = transaction
+                .try_into()
+                .map_err(|_| ErrorObject::owned(INVALID_PARAMS_CODE, "Failed to convert transaction", None::<()>))?;
```

Comment:
Here I think I don't understand why we need to put the Error Object with the error code directly like that? 

It seemed to me that in the rest of the codebase we used a different approach with error generation such as `EthApiError::EthereumDataFormat`. Then as we have the implementation of the conversion between this type of error and `EthRpcErrorCode`, the conversion to the correct error code is done automatically right @greged93 ? But maybe I missed something here

Reply:
I might have missed something but I had seen the other errors defined in the codebase, but they all seemed very specific to me, such as errors in the transaction fields, or EVM/Cairo VM errors. I thought they didn't fit well in this case, but I might have missed something regarding the errors. 

To avoid inserting the error code directly, can I add a more generic error in `KakarotError`. What do you think?